azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/module_client.py

```diff
@@ -4,7 +4,19 @@
 # --------------------------------------------------------------------------------------------
 
 from .internal_client import InternalClient
+from .message import Message
 
 
 class ModuleClient(InternalClient):
-    pass
+    def send_to_output(self, message, output_name):
+        """
+        Sends an event/message to the given module output.
+        This is an outgoing events and  means to be "output events"
+        :param output_name: Name of the output to send the event to.
+        :param message: message to send to the given output
+        """
+        if not isinstance(message, Message):
+            message = Message(message)
+
+        message.output_name = output_name
+        self.send_event(message)
```

Comment:
>self.send_event(message) [](start = 8, length = 24)

I think this is a great reuse of code and I like how you leveraged the existing code.  I recommend finding a way to do this without calling self.send_event, maybe calling transport.send_event from inside here, or maybe make a private function like "_send_event_common" that is called by both public functions  I'm not sure if there's a coding "principle" that speaks against public functions that call other public functions, but this smells like bad layering.    It's important to have one set of functions that only gets called by customers (the public API) and a different set of functions that only get called by internal code (private functions).  This doesn't matter too much now because send_to_output does a little bit more than send_event, but it will matter more later if send_event needs to do something that send_to_output doesn't.

Reply:
Comment #2: all other functions (so far) have a 1:1 mapping from client function name to transport function name.  I don't think that's a requirement, but it seems strange that client.connect calls transport.connect, client.disconnect calls transport.disconnect, client.send_event calls transport.send_event, and (the one exception) client.send_to_output calls transport.send_event.  By having the transport functions be exactly the same name as the client functions, we're implying a relationship, but this code breaks that relationship.  If the transport function handles telemetry and output events both, that OK, but calling it send_event implies that it only handles telemetry.  Maybe we should rename the transport function since the 1:1 mapping doesn't exist here.  

---
In reply to: [246254214](https://github.com/Azure/azure-iot-sdk-python-preview/pull/35#discussion_r246254214) [](ancestors = 246254214)

Reply:
The reason I utilized the send event code is because of the threading code. The only thing that send to output does now is setting the output name. It does nothing else as of now. But true it may do something in the future.
For comment 2 s it is a correct point that the transport should have 1:1 mapping. The trigger called can be same. So if the transport has another end point called send output , we don't need the common function.


Also I am not sure if the AbstractTransport should have that send output as an abstract method send output

---
In reply to: [246257157](https://github.com/Azure/azure-iot-sdk-python-preview/pull/35#discussion_r246257157) [](ancestors = 246257157,246254214)